src/services/transferUpload.ts

```diff
@@ -0,0 +1,186 @@
+        uploadData,
+        reservationToken: reservation.token,
+        shouldCheckUploadedFileSize: false,
+        retryCount: 0,
+        pushProgress,
+      });
+      completedUploadSize += uploadData.reservedFile.size;
+    }
+
+    onReservationConfirm();
+    const result = await this.skpApi.confirmReservation(reservation.token);
+    await delay(2);
+    return result;
+  }
+
+  private async initiateUpload(props: InitiateUploadProp): Promise<void> {
+    if (props.shouldCheckUploadedFileSize) {
+      try {
+        props.uploadData.startByte = await this.skpApi.fetchUploadedFileSize(
+          props.uploadData.reservedFile.objectId,
+          props.reservationToken
+        );
+      } catch (error) {
+        if (props.retryCount >= constants.maxUploadRetries) throw error;
+        if (axios.isAxiosError(error) && error.response?.status === 404) {
+          props.uploadData.startByte = 0;
+        } else {
+          return await this.retryUpload(props);
+        }
+      }
+    }
+
+    props.uploadData.endByte = Math.min(
+      props.uploadData.startByte + props.currentChunkSize,
+      props.uploadData.reservedFile.size
+    );
+
+    props.uploadData.formData = await this.createUploadFormData(
+      props.uploadData,
+      props.reservationToken
+    );
+
+    try {
+      const chunkUploadStartTime = Date.now();
+      const response: UploadInfo = await this.skpApi.uploadFileChunk({
+        startByte: props.uploadData.startByte,
+        endByte: props.uploadData.endByte,
+        formData: props.uploadData.formData,
+        totalBytes: props.uploadData.reservedFile.size,
+        onUploadProgress: (event) =>
+          props.pushProgress(event.loaded, props.uploadData.startByte),
+      });
+
+      props.retryCount = 0;
+      const duration = Math.floor((Date.now() - chunkUploadStartTime) / 1000);
+      const newChunkSize = this.calculateNewChunkSize(
+        duration,
+        props.currentChunkSize
+      );
+
+      if (response.size < props.uploadData.reservedFile.size) {
+        props.uploadData.startByte = response.size;
+
+        return await this.initiateUpload({
+          ...props,
+          currentChunkSize: newChunkSize,
+          shouldCheckUploadedFileSize: false,
+        });
+      }
+    } catch (error) {
+      if (props.retryCount >= constants.maxUploadRetries) throw error;
+      return await this.retryUpload(props);
+    }
+  }
+
+  private async retryUpload(props: InitiateUploadProp): Promise<void> {
+    props.retryCount++;
+    const wait = Math.min(
+      2 ** props.retryCount,
+      constants.maxDelayBetweenRetriesSec
+    );
+    await delay(wait);
+    props.uploadData.startByte = 0;
+    props.shouldCheckUploadedFileSize = true;
+    return await this.initiateUpload(props);
+  }
+
+  private async createUploadFormData(
+    uploadData: UploadData,
+    token: string
+  ): Promise<FormData> {
+    const blob = await fs.openAsBlob(uploadData.filePath);
```

Comment:
question: does this work with very large files?

Reply:
I'm not sure tbh, I updated the code to use the readstream of the file instead. I have found this to be the recommended way of doing uploads. RAM usage is still relatively high for large file uploads but there seems to be little we can do to improve this further. Somewhere upstream buffering still seems to be happening I guess